### ISSUE TITLE
Possibility to port zstd decompression to ARM based MCU?

### ISSUE THREAD

**Issue by JiapengLi** [facebook/zstd#1612]
I happened to know ZSTD library when I did research about LZ4, and made comparison. The performance is very surprised. Thank you for the great project!

My use case is firmware delta update on resources limited MCU. The concept is to use old firmware as dictionary and send delta file (*.zst) over low speed wireless communication, MCU receives the firmware and extract it to get new firmware to finish firmware upgrade.  Thus the delta file the small the better. And at the same time, decompress can't be too complex, because decompress is executed on MCU. Firstly I evaluate jojoDiff and LZ4, the performance is similar (LZ4 is better if firmware size is less than 64KB). Both jojoDiff and LZ4 destination file format is simple and it is easy for MCU to do decompress. I read ZSTD file format, it looks more complex than LZ4's and jojodiff's file format. **Is it possible to run ZSTD decompress on MCU, what is the minimum memory and program cost? Thank you!** 

The MCU I use has 20KB RAM and 128KB ROM.

Comparison between JojoDiff and ZSTD-19

Old File | New File | Changed | Ratio | JojoDiff | JojoDiff | zstd-19 | zstd-19
-- | -- | -- | -- | -- | -- | -- | --
117388 | 121291 | 105745 | 90.1% | 37800 | 31.2% | 17872 | 14.7%
65396 | 65496 | 62620 | 95.8% | 44291 | 67.6% | 26240 | 40.1%
512 | 512 | 18 | 3.5% | 59 | 11.5% | 40 | 7.8%

**Comment by Cyan4973** [comment]
I believe this is at the limit of what zstd can do.
It's already possible to reduce the decoder to ~12 KB, which is already high by MCU standard.
More importantly, the RAM budget is not bounded by the decoder size, which, in its current incarnation, requires much more buffering (about ~150KB). It's possible to design the decoder differently, so that it consumes much less RAM. I believe the lower bound is somewhere in the ~7 KB range. But creating such a version to fit into this MCU budget will be a significant effort.

So basically, current reference implementation of `zstd` is not meant to fit into this memory budget.
Making a version specifically designed for the restrictions of this platform is possible, but is also a significant investment.

**Comment by JiapengLi** [comment]
@Cyan4973 Thank you for your prompt reply. I understand the situation.
